submissions/salsita/src/reducers/scrollingReducer.js

```diff
@@ -0,0 +1,145 @@
+      mutableReduction.setIn(['appState', 'disabledDown'], true);
+    } else {
+      mutableReduction.setIn(['appState', 'disabledUp'], false);
+      mutableReduction.setIn(['appState', 'disabledDown'], false);
+    }
+  });
+};
+
+export const scrolledDown = reduction => reduction.withMutations(mutableReduction => {
+  if (!mutableReduction.getIn(['appState', 'disabledDown']) && !mutableReduction.getIn(['appState', 'sithSpotted'])) {
+    // Just a flag that we are scrolling down
+    mutableReduction.setIn(['appState', 'scrollingDown'], true);
+
+    // Let's take the first N Siths which went out of the screen
+    // and find out whether they are still loading.
+    // This implementation should probably take List of Siths, not single Sith.
+    const cancelableSith = mutableReduction
+      .getIn(['appState', 'siths'])
+      .take(Configuration.PAGINATION_STEP)
+      .find(sith => sith.get('state') === SithStates.LOADING);
+
+    // If there is still a Sith which is loading, cancel its request.
+    if (cancelableSith) {
+      mutableReduction.update('effects', effects => effects.push(buildMessage(Effects.API_CANCEL_REQUEST, cancelableSith.get('correlationId'))));
+    }
+
+    // Move all the visible Siths up. We should skip those which got out of screen.
+    mutableReduction
+      .getIn(['appState', 'siths'])
+      .skip(Configuration.PAGINATION_STEP)
+      .forEach((sith, index) => mutableReduction.setIn(['appState', 'siths', index], sith));
+
+    const lastSithOnScreenIndex = Configuration.SITH_SLOTS - Configuration.PAGINATION_STEP - 1;
+    const nextSithToLoadIndex = lastSithOnScreenIndex + 1;
+
+    // Let's get an apprentice of the last Sith on the screen (because we are scrolling down)
+    const apprenticeId = mutableReduction.getIn(['appState', 'siths', lastSithOnScreenIndex, 'apprentice']);
+
+    // Current Sith might not have next apprentice, therefore we are on the end of the list
+    if (apprenticeId) {
+      // The next Sith on screen should start loading (it's apprentice of the last Sith on screen)
+      mutableReduction
+        .setIn(['appState', 'siths', nextSithToLoadIndex], new SithRecord({
+          id: apprenticeId,
+          state: SithStates.LOADING
+        }))
+        .update('effects', effects => effects.push(buildMessage(Effects.API_LOAD_SITH, {
+          sithId: apprenticeId,
+          scrollingDown: true
+        })));
+    } else {
+      mutableReduction.setIn(['appState', 'siths', nextSithToLoadIndex], new SithRecord());
+    }
+
+    // Let's just clean all the Siths bellow (pagination step)
+    mutableReduction.updateIn(['appState', 'siths'], siths => siths.map((sith, index) => {
+      if (index > Configuration.SITH_SLOTS - Configuration.PAGINATION_STEP) {
+        return new SithRecord();
+      } else {
+        return sith;
+      }
+    }));
+  }
+});
+
+export const scrolledUp = reduction => reduction.withMutations(mutableReduction => {
```

Comment:
I hate all this duplicated code. I think you need to find an elegant way to move as much code as possible from `scrolledUp` and `scrolledDown` into a shared function.